prime-dotnet-webapi/Services/AuthorizedUserService.cs

```diff
@@ -0,0 +1,137 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using AutoMapper;
+using AutoMapper.QueryableExtensions;
+using Microsoft.AspNetCore.Http;
+using Microsoft.EntityFrameworkCore;
+using Prime.Extensions;
+using Prime.Models;
+using Prime.ViewModels.Parties;
+
+namespace Prime.Services
+{
+    public class AuthorizedUserService : BaseService, IAuthorizedUserService
+    {
+        private readonly IMapper _mapper;
+        private readonly IPartyService _partyService;
+
+        public AuthorizedUserService(
+            ApiDbContext context,
+            IMapper mapper,
+            IHttpContextAccessor httpContext,
+            IPartyService partyService)
+            : base(context, httpContext)
+        {
+            _mapper = mapper;
+            _partyService = partyService;
+        }
+
+        public async Task<bool> AuthorizedUserExistsAsync(int authorizedUserId)
+        {
+            return await _context.AuthorizedUsers
+                .AsNoTracking()
+                .AnyAsync(au => au.Id == authorizedUserId);
+        }
+
+        public async Task<bool> AuthorizedUserExistsForUserIdAsync(Guid userId)
+        {
+            return await _partyService.PartyExistsForUserIdAsync(userId, PartyType.AuthorizedUser);
+        }
+
+        public async Task<AuthorizedUserViewModel> GetAuthorizedUserAsync(int authorizedUserId)
+        {
+            return await GetBaseAuthorizedUserQuery()
+                .Where(au => au.Id == authorizedUserId)
+                .ProjectTo<AuthorizedUserViewModel>(_mapper.ConfigurationProvider)
+                .SingleOrDefaultAsync();
+        }
+
+        public async Task<AuthorizedUserViewModel> GetAuthorizedUserForUserIdAsync(Guid userId)
+        {
+            return await GetBaseAuthorizedUserQuery()
+                .Where(au => au.Party.UserId == userId)
+                .ProjectTo<AuthorizedUserViewModel>(_mapper.ConfigurationProvider)
+                .SingleOrDefaultAsync();
+        }
+
+        public async Task<int> CreateOrUpdateAuthorizedUserAsync(AuthorizedUserChangeModel changeModel, ClaimsPrincipal user)
+        {
+            var authorizedUser = await GetBaseAuthorizedUserQuery()
+                .SingleOrDefaultAsync(au => au.Party.UserId == user.GetPrimeUserId());
+
+            if (authorizedUser == null)
+            {
+                var party = new Party
+                {
+                    Addresses = new List<PartyAddress>()
+                };
+                _context.Parties.Add(party);
```

Comment:
This will throw if the User already has a Party in the database but isn't already an AuthorizedUser. This will try to add another Party to the DB with a duplicate UserId

Reply:
As an aside: if you are creating multiple entities at the same time, you shouldn't need to add them to all the DB sets individually; EF tracks the whole object graph. This block of code could be simplified to 
```C#
authorizedUser = new AuthorizedUser
{
    Party = new Party
    {
        Addresses = new List<PartyAddress>()
    },
    Status = AccessStatusType.UnderReview
};
_context.AuthorizedUsers.Add(authorizedUser);
```

Reply:
@james-hollinger I thought mine was a small improvement over the `PartyService` version, but love that it is fully tracked so I updated the below from `PartyService`:

![image](https://user-images.githubusercontent.com/599498/118195511-32c98780-b400-11eb-9164-bf6fa28784a1.png)

to:

![image](https://user-images.githubusercontent.com/599498/118195636-702e1500-b400-11eb-881e-cb41ce8492dc.png)